Addressed SaferCPP RetainPtr adopt checker issues in WebKit

geoffreygaren opened the pull request:
#### 717849ac7d5a328b011e686499f04194b72809c7
<pre>
Addressed SaferCPP RetainPtr adopt checker issues in WebKit
<a href="https://bugs.webkit.org/show_bug.cgi?id=300367">https://bugs.webkit.org/show_bug.cgi?id=300367</a>
&lt;<a href="https://rdar.apple.com/problem/162169383">rdar://problem/162169383</a>&gt;

Reviewed by David Kilzer.

Mostly skipped some odd cases.

Updated some edge case code to use idiomatic adopt().

Canonical link: <a href="https://commits.webkit.org/301233@main">https://commits.webkit.org/301233@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1d169b0ab55a98fa61d5b82758abc7fa01dc503c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/125317 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/44982 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/35723 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/132166 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/77185 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 ios-apple 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/45670 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/123/builds/53547 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/95410 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/77185 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/128271 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/36476 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/112066 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/75949 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | ⏳ 🛠 vision-apple 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/35375 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/75643 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/106244 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/30461 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/134852 "Built successfully") | | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/128/builds/52120 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/123/builds/53547 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/103883 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/52558 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/108288 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/103644 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/49008 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/27299 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/49230 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/19623 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/52015 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/57795 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/51372 "Built successfully") | | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/54728 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/53066 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

ddkilzer commented on Source/WebKit/Shared/API/c/cf/WKErrorCF.cpp:
```diff
@@ -37,6 +37,6 @@ WKErrorRef WKErrorCreateWithCFError(CFErrorRef cfError)
 CFErrorRef WKErrorCopyCFError(CFAllocatorRef alloc, WKErrorRef error)
 {
     WTF::RetainPtr<CFErrorRef> cfError = WebKit::toImpl(error)->platformError().cfError();
-    return cfError.leakRef();
+    // Leaking is correct because our API is a copy.
+    SUPPRESS_RETAINPTR_CTOR_ADOPT return cfError.leakRef();
 }
```
Why doesn't the existence of `CF_RETURNS_RETAINED` on the function declaration present this?

```
WK_EXPORT CFErrorRef WKErrorCopyCFError(CFAllocatorRef alloc, WKErrorRef error) CF_RETURNS_RETAINED;
```

The checker should be able to reason about this.

ddkilzer commented on Source/WebKit/Shared/Cocoa/WKNSString.mm:
```diff
@@ -37,7 +37,8 @@ @implementation WKNSString
 - (NSObject *)_web_createTarget
 {
     String string = RefPtr { downcast<API::String>(&self._apiObject) }->string();
-    return bridge_cast(string.createCFString().leakRef());
+    // Leaking is correct because our API is a copy.
+    SUPPRESS_RETAINPTR_CTOR_ADOPT return bridge_cast(string.createCFString().leakRef());
 }
```
Similarly, this method is declared as `NS_RETURNS_RETAINED`, so the checker should be able to reason about this.

ddkilzer commented on Source/WebKit/Shared/API/Cocoa/_WKRemoteObjectRegistry.mm:
```diff
@@ -298,7 +298,8 @@ - (void)_invokeMethod:(const WebKit::RemoteObjectInvocation&)remoteObjectInvocat
         RefPtr<ReplyBlockCallChecker> checker = ReplyBlockCallChecker::create(self, replyID);
-        auto replyBlock = adoptNS(__NSMakeSpecialForwardingCaptureBlock([wireBlockSignature UTF8String], [interface, remoteObjectRegistry, replyID, checker](NSInvocation *invocation) {
+        // Despite the name, __NSMakeSpecialForwardingCaptureBlock returns +1
+        SUPPRESS_RETAINPTR_CTOR_ADOPT auto replyBlock = adoptNS(__NSMakeSpecialForwardingCaptureBlock([wireBlockSignature UTF8String], [interface, remoteObjectRegistry, replyID, checker](NSInvocation *invocation) {
```
Do we have a radar asking for `NS_RETURNS_RETAINED` to be added to `__NSMakeSpecialForwardingCaptureBlock`?

ddkilzer commented on Source/WebKit/Shared/API/c/cg/WKImageCG.cpp:
```diff
@@ -46,7 +46,8 @@ CGImageRef WKImageCreateCGImage(WKImageRef imageRef)
         return nullptr;
 
     auto platformImage = nativeImage->platformImage();
-    return platformImage.leakRef();
+    // Leaking is correct because our API is a create.
+    SUPPRESS_RETAINPTR_CTOR_ADOPT return platformImage.leakRef();
 }
```
This function's declaration also has `CF_RETURNS_RETAINED`.

ddkilzer commented on Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreePropertyApplier.mm:
```diff
@@ -312,7 +312,7 @@ static void applyVisualStylingToLayer(CALayer *layer, const AppleVisualEffectDat
 
     // Despite the name, MTVisualStylingCreateDictionaryRepresentation returns an autoreleased object.
-    RetainPtr visualStylingDescription = PAL::softLink_CoreMaterial_MTVisualStylingCreateDictionaryRepresentation(recipe.get(), materialVisualStyleCategoryForAppleVisualEffect(effectData.effect), materialVisualStyleForAppleVisualEffect(effectData.effect), nil);
+    SUPPRESS_RETAINPTR_CTOR_ADOPT RetainPtr visualStylingDescription = PAL::softLink_CoreMaterial_MTVisualStylingCreateDictionaryRepresentation(recipe.get(), materialVisualStyleCategoryForAppleVisualEffect(effectData.effect), materialVisualStyleForAppleVisualEffect(effectData.effect), nil);
```
Are we missing an `NS_RETURNS_RETAINED` on the softlink definition?

geoffreygaren replied:
We actually need NS_RETURNS_AUTORELEASED. I would add it now, but we can't test yet because our analyzer doesn't support it yet.

ddkilzer commented on Source/WebKit/UIProcess/API/Cocoa/WKBackForwardListItem.mm:
```diff
@@ -77,8 +77,10 @@ - (NSURL *)initialURL
 
 - (CGImageRef)_copySnapshotForTesting
```
The method declaration in `Source/WebKit/UIProcess/API/Cocoa/WKBackForwardListItemPrivate.h` needs `CF_RETURNS_RETAINED`.

```
- (CGImageRef)_copySnapshotForTesting WK_API_AVAILABLE(macos(10.12.3), ios(10.3));
```

ddkilzer commented on Source/WebKit/UIProcess/API/Cocoa/WKBackForwardListItem.mm:
```diff
@@ -77,8 +77,10 @@ - (NSURL *)initialURL
 {
-    if (RefPtr snapshot = _item->snapshot())
-        return snapshot->asImageForTesting().leakRef();
+    if (RefPtr snapshot = _item->snapshot()) {
+        // Leaking is correct because our API is a copy.
+        SUPPRESS_RETAINPTR_CTOR_ADOPT return snapshot->asImageForTesting().leakRef();
+    }
```
Checker should be able to understand `CF_RETURNS_RETAINED` on the method declaration.

ddkilzer reviewed (approved):
Please:
- File a radar for `__NSMakeSpecialForwardingCaptureBlock()` (and add comment)
- Add `CF_RETURNS_RETAINED` to `_copySnapshotForTesting` method
- File a radar to make RetainPtr check understand leaks in `return` statements when `CF_RETURNS_RETAINED` or `NS_RETURNS_RETAINED` is used on the function/method declaration

r=me with these steps taken

geoffreygaren committed:
Addressed SaferCPP RetainPtr adopt checker issues in WebKit
https://bugs.webkit.org/show_bug.cgi?id=300367
<rdar://problem/162169383>

Reviewed by David Kilzer.

Mostly skipped some odd cases.

Updated some edge case code to use idiomatic adopt().

Canonical link: https://commits.webkit.org/301233@main
